pynbody/halo.py

```diff
@@ -875,24 +897,120 @@ def __init__(self, sim, make_grp=None, dummy=False, use_iord=None, ahf_basename=
         if config_parser.getboolean('AHFCatalogue', 'AutoPid'):
             sim['pid'] = np.arange(0, len(sim))
 
+        if write_fpos is not None:
+            if not os.path.exists(self._ahfBasename + 'fpos'):
+                self._write_fpos()
+
         logger.info("AHFCatalogue loaded")
 
+    def __getitem__(self,item):
+        """
+        get the appropriate halo if dosort is on
+        """
+        if self._dosort is not None:
+            i = self._sorted_indices[item-1]
+        else:
+            i = item
+        return super(AHFCatalogue,self).__getitem__(i)
+
     def make_grp(self, name='grp'):
         """
         Creates a 'grp' array which labels each particle according to
         its parent halo.
         """
         self.base[name] = self.get_group_array()
 
-    def get_group_array(self, top_level=False):
-        ar = np.zeros(len(self.base), dtype=int)
+    def _write_fpos(self):
+        f = open(self._ahfBasename + 'fpos','w')
+        for i in range(self._nhalos):
+            if i < self._nhalos - 1:
+                f.write(str(self._halos[i+1].properties['fstart'])+'\n')
+            else:
+                f.write(str(self._halos[i+1].properties['fstart']))
+        f.close()
+
+
+    def get_group_array(self, top_level=False, family=None):
```

Comment:
I think `get_group_array` is a bit dangerous - quite obscure, and hard-codes a whole load of notions about the different families (e.g. the order they appear in the file)

The correct, and simpler, way to do this would be to use `self.base._get_family_slice(family)`, which returns a slice, so for example `self.base._get_family_slice(family).start` tells you where the first index is. Then you don't need to do it case-by-case or expect a certain ordering for the families.